core/src/main/java/io/grpc/internal/OutboundState.java

```diff
@@ -0,0 +1,294 @@
+    checkNotNull(message);
+    phase(Phase.MESSAGE);
+    if (!framer.isClosed()) {
+      framer.writePayload(message);
+    }
+  }
+
+  /**
+   * Request to flush any buffered messages. Must be called form the channel thread context.
+   */
+  void flush() {
+    if (!framer.isClosed()) {
+      framer.flush();
+    }
+  }
+
+  /**
+   * If {@code true}, indicates that the transport is capable of sending additional messages without
+   * requiring excessive buffering internally. Otherwise, {@link Handler#onReady()} will be
+   * called when it turns {@code true}.
+   *
+   * <p>This is just a suggestion and the application is free to ignore it, however doing so may
+   * result in excessive buffering within the transport.
+   */
+  boolean isReady() {
+    if (phase != Phase.STATUS) {
+      synchronized (onReadyLock) {
+        return numSentBytesQueued < ONREADY_THRESHOLD;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Closes this outbound context. Must be called from the application thread context.
+   */
+  void close(Status status, boolean send) {
+    checkNotNull(status, "status");
+    if (phase(Phase.STATUS) == Phase.STATUS) {
+      // Already closed.
+      return;
+    }
+
+    if (!send) {
+      // Close the framer, but do not flush any remaining data.
+      framer.dispose();
+      return;
+    }
+
+    if (server) {
+      // TODO(louiscryan): Remove
+      if (stashedTrailers == null) {
+        stashedTrailers = new Metadata();
+      }
+      writeStatusToTrailers(status);
+    }
+
+    // Close the framer and flush any remaining data. For server-side, this will trigger
+    // the sending of trailers.
+    framer.close();
+  }
+
+  /**
+   * Notifies this context that the given number of bytes has actually been sent to the remote
+   * endpoint. May call {@link Handler#onReady()} if appropriate. This should be called
+   * from the transport thread.
+   *
+   * @param numBytes the number of bytes that were sent by the transport.
+   */
+  void bytesSent(int numBytes) {
+    boolean doNotify;
+    synchronized (onReadyLock) {
+      boolean belowThresholdBefore = numSentBytesQueued < ONREADY_THRESHOLD;
+      numSentBytesQueued -= numBytes;
+      boolean belowThresholdAfter = numSentBytesQueued < ONREADY_THRESHOLD;
+      doNotify = !belowThresholdBefore && belowThresholdAfter;
+    }
+    if (doNotify) {
+      // Notify the handler that we're ready to accept more data.
+      handler.onReady();
+    }
+  }
+
+  /**
+   * Updates the number of bytes are being queued for sending to the remote endpoint.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  private void bytesQueued(int numBytes) {
+    if (numBytes > 0) {
+      synchronized (onReadyLock) {
+        numSentBytesQueued += numBytes;
+      }
+    }
+  }
+
+  /**
+   * Handles the output of the {@link MessageFramer}.
+   */
+  private void onMessageFrame(final WritableBuffer frame, final boolean endOfStream,
```

Comment:
nit: since there aren't any anonymous classes, consider not including final.  

final on local variables makes me look for anon classes, which is surprising to me when there aren't any.